File: app/src/main/java/com/arjanvlek/oxygenupdater/MainActivity.java
package com.arjanvlek.oxygenupdater;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.v7.app.ActionBar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.app.FragmentPagerAdapter;
import android.os.Bundle;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.Toast;

import com.arjanvlek.oxygenupdater.Support.NetworkConnectionManager;
import com.arjanvlek.oxygenupdater.Support.SettingsManager;
import com.arjanvlek.oxygenupdater.Support.SupportedDeviceCallback;
import com.arjanvlek.oxygenupdater.Support.SupportedDeviceManager;
import com.arjanvlek.oxygenupdater.views.FAQActivity;
import com.arjanvlek.oxygenupdater.views.HelpActivity;
import com.arjanvlek.oxygenupdater.views.MessageDialog;
import com.arjanvlek.oxygenupdater.views.SettingsActivity;
import com.arjanvlek.oxygenupdater.views.AboutActivity;
import com.arjanvlek.oxygenupdater.views.DeviceInformationFragment;
import com.arjanvlek.oxygenupdater.views.TutorialActivity;
import com.arjanvlek.oxygenupdater.views.UpdateInformationFragment;
import com.arjanvlek.oxygenupdater.views.UpdateInstallationGuideActivity;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GooglePlayServicesUtil;

import static com.arjanvlek.oxygenupdater.Support.SettingsManager.*;


@SuppressWarnings("deprecation")
public class MainActivity extends AppCompatActivity implements ActionBar.TabListener, SupportedDeviceCallback {

    private ViewPager mViewPager;
    private SettingsManager settingsManager;
    private NetworkConnectionManager networkConnectionManager;

    // Used for Google Play Services check
    private final static int PLAY_SERVICES_RESOLUTION_REQUEST = 9000;

    // Permissions constants
    public final static String DOWNLOAD_PERMISSION = "android.permission.WRITE_EXTERNAL_STORAGE";
    public final static int PERMISSION_REQUEST_CODE = 200;

    private String device = "";
    private long deviceId = 0L;
    private String updateMethod = "";
    private long updateMethodId = 0L;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_activity);
        Context context = getApplicationContext();
        settingsManager = new SettingsManager(context);
        networkConnectionManager = new NetworkConnectionManager(context);

        if(!settingsManager.getBooleanPreference(SettingsManager.PROPERTY_IGNORE_UNSUPPORTED_DEVICE_WARNINGS) && networkConnectionManager.checkNetworkConnection()) {
            SupportedDeviceManager supportedDeviceManager = new SupportedDeviceManager(this, ((ApplicationContext)getApplication()));
            supportedDeviceManager.execute();
        }

        //Fetch currently selected device and update method
        device = settingsManager.getPreference(PROPERTY_DEVICE);
        deviceId = settingsManager.getLongPreference(PROPERTY_DEVICE_ID);
        updateMethod = settingsManager.getPreference(PROPERTY_UPDATE_METHOD);
        updateMethodId = settingsManager.getLongPreference(PROPERTY_UPDATE_METHOD_ID);

        // Set up the action bar.
        final ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_TABS);
        }
        setTitle(getString(R.string.app_name));

        SectionsPagerAdapter mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());

        mViewPager = (ViewPager) findViewById(R.id.mainActivityPager);

        if(mViewPager != null) {
            mViewPager.setAdapter(mSectionsPagerAdapter);
            mViewPager.setOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
                @Override
                public void onPageSelected(int position) {
                    if (actionBar != null) {
                        actionBar.setSelectedNavigationItem(position);
                    }
                }
            });
        }



        // For each of the sections in the app, add a tab to the action bar.
        for (int i = 0; i < mSectionsPagerAdapter.getCount(); i++) {
            // Creates a tab with text corresponding to the page title defined by
            // the adapter.
            //noinspection ConstantConditions
            actionBar.addTab(
                    actionBar.newTab()
                            .setText(mSectionsPagerAdapter.getPageTitle(i))
                            .setTabListener(this));
        }

    }

    @Override
    public void onStart() {
        super.onStart();
        // Check if Google Play services are installed on the device
        if (checkPlayServices()) {
            // Check if a device and update method have been set
            if (settingsManager.checkIfDeviceIsSet()) {
                //Check if app needs to register for push notifications (like after device type change etc.)
                if(device != null && updateMethod != null) {
                    if (!settingsManager.checkIfRegistrationIsValid(deviceId, updateMethodId) || settingsManager.checkIfRegistrationHasFailed() && networkConnectionManager.checkNetworkConnection()) {
                        registerInBackground();
                    }
                }
            }

            // Mark the welcome tutorial as finished if the user is moving from older app version. This is checked by either having stored update information for offline viewing, or if the last update checked date is set (if user always had up to date system and never viewed update information before).
            if(!settingsManager.getBooleanPreference(PROPERTY_SETUP_DONE) && (settingsManager.checkIfCacheIsAvailable() || settingsManager.containsPreference(PROPERTY_UPDATE_CHECKED_DATE))) {
                settingsManager.saveBooleanPreference(PROPERTY_SETUP_DONE, true);
            }

            // Show the welcome tutorial if the app needs to be set up.
            if(!settingsManager.getBooleanPreference(PROPERTY_SETUP_DONE)) {
                if(networkConnectionManager.checkNetworkConnection()) {
                    Tutorial();
                } else {
                    showNetworkError();
                }
            }

        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    private void showNetworkError() {
        MessageDialog errorDialog = new MessageDialog()
                .setTitle(getString(R.string.error_app_requires_network_connection))
                .setMessage(getString(R.string.error_app_requires_network_connection_message))
                .setNegativeButtonText(getString(R.string.download_error_close))
                .setClosable(false);
        errorDialog.show(getSupportFragmentManager(), "NetworkError");
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handles action bar item clicks.
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            Settings();
            return true;
        }
        if (id == R.id.action_about) {
            About();
            return true;
        }

        if (id == R.id.action_help) {
            Help();
            return true;
        }

        if (id == R.id.action_faq) {
            FAQ();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * Action when clicked on a tab.
     * @param tab Tab which is selected
     * @param fragmentTransaction Android Fragment Transaction, unused here.
     */
    @Override
    public void onTabSelected(ActionBar.Tab tab, FragmentTransaction fragmentTransaction) {
        // When the given tab is selected, switch to the corresponding page in
        // the ViewPager.
        mViewPager.setCurrentItem(tab.getPosition());
    }

    @Override
    public void onTabUnselected(ActionBar.Tab tab, FragmentTransaction fragmentTransaction) {
    }

    @Override
    public void onTabReselected(ActionBar.Tab tab, FragmentTransaction fragmentTransaction) {
    }

    @Override
    public void displayUnsupportedMessage(boolean deviceIsSupported) {

        if(!settingsManager.getBooleanPreference(PROPERTY_IGNORE_UNSUPPORTED_DEVICE_WARNINGS) && !deviceIsSupported) {
            View checkBoxView = View.inflate(MainActivity.this, R.layout.alert_dialog_checkbox, null);
            final CheckBox checkBox = (CheckBox) checkBoxView.findViewById(R.id.unsupported_device_warning_checkbox);

            AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
            builder.setView(checkBoxView);
            builder.setTitle(getString(R.string.unsupported_device_warning_title));
            builder.setMessage(getString(R.string.unsupported_device_warning_message));

            builder.setPositiveButton(getString(R.string.download_error_close), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    SettingsManager settingsManager = new SettingsManager(getApplicationContext());
                    settingsManager.saveBooleanPreference(SettingsManager.PROPERTY_IGNORE_UNSUPPORTED_DEVICE_WARNINGS, checkBox.isChecked());
                    dialog.dismiss();
                }
            });
            builder.show();
        }
    }


    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentPagerAdapter {

        public SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            // Return a FragmentBuilder (defined as a static inner class below).
            return FragmentBuilder.newInstance(position + 1);
        }

        @Override
        public int getCount() {
            // Show 2 total pages.
            return 2;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            boolean MorHigher = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
            switch (position) {
                case 0:
                    return MorHigher ? getString(R.string.update_information_header_short) : getString(R.string.update_information_header);
                case 1:
                    return MorHigher ? getString(R.string.device_information_header_short) : getString(R.string.device_information_header);
            }
            return null;
        }
    }

    /**
     * An inner class that constructs the fragments used in this application.
     */
    public static class FragmentBuilder {

        /**
         * Returns a new instance of this fragment for the given section
         * number.
         */
        public static Fragment newInstance(int sectionNumber) {
            if (sectionNumber == 1) {
                return new UpdateInformationFragment();
            }
            if (sectionNumber == 2) {
                return new DeviceInformationFragment();
            }
            return null;
        }
    }

    /**
     * Opens the settings page.
     */
    private void Settings() {
        Intent i = new Intent(this, SettingsActivity.class);
        startActivity(i);
    }

    /**
     * Opens the welcome tutorial.
     */
    private void Tutorial() {
        Intent i = new Intent(this, TutorialActivity.class);
        startActivity(i);
    }

    /**
     * Opens the about page.
     */
    private void About() {
        Intent i = new Intent(this, AboutActivity.class);
        startActivity(i);
    }

    /**
     * Opens the help page.
     */
    private void Help() {
        Intent i = new Intent(this, HelpActivity.class);
        startActivity(i);
    }

    /**
     * Opens the faq page.
     */
    private void FAQ() {
        Intent i = new Intent(this, FAQActivity.class);
        startActivity(i);
    }


    /**
     * Opens the update instructions page.
     * @param v View (button onclick from XML).
     */
    public void UpdateInstructions(View v) {
        Intent i = new Intent(this, UpdateInstallationGuideActivity.class);
        startActivity(i);
    }


    /**
     * Checks if the Google Play Services are installed on the device.
     * @return Returns if the Google Play Services are installed.
     */
    private boolean checkPlayServices() {
        int resultCode = GooglePlayServicesUtil.isGooglePlayServicesAvailable(this);
        if (resultCode != ConnectionResult.SUCCESS) {
            if (GooglePlayServicesUtil.isUserRecoverableError(resultCode)) {
                GooglePlayServicesUtil.getErrorDialog(resultCode, this,
                        PLAY_SERVICES_RESOLUTION_REQUEST).show();
            } else {
                finish();
            }
            return false;
        }
        return true;
    }

    /**
     * Registers the application with GCM servers asynchronously.
     * Stores the registration ID and app versionCode in the application's
     * shared preferences.
     */
    private void registerInBackground() {
        Intent intent = new Intent(this,GcmRegistrationIntentService.class);
        startService(intent);
    }

    // New Android 6.0 permissions methods

    public void requestDownloadPermissions() {
        if(Build.VERSION.SDK_INT>= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{DOWNLOAD_PERMISSION}, PERMISSION_REQUEST_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int  permsRequestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (permsRequestCode) {
            case PERMISSION_REQUEST_CODE:
                try {
                    if(grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                        Toast.makeText(this, getString(R.string.download_permissions), Toast.LENGTH_LONG).show();
                    }
                } catch (Exception ignored) {

                }
        }
    }

    public boolean hasDownloadPermissions() {
        //noinspection SimplifiableIfStatement
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return (checkSelfPermission(DOWNLOAD_PERMISSION) == PackageManager.PERMISSION_GRANTED);
        } else {
            return true;
        }
    }
}

File: app/src/main/java/com/arjanvlek/oxygenupdater/views/MainActivity.kt
package com.arjanvlek.oxygenupdater.views

import android.annotation.TargetApi
import android.app.AlertDialog
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import androidx.core.view.children
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.arjanvlek.oxygenupdater.ActivityLauncher
import com.arjanvlek.oxygenupdater.ApplicationData
import com.arjanvlek.oxygenupdater.ApplicationData.Companion.buildAdRequest
import com.arjanvlek.oxygenupdater.R
import com.arjanvlek.oxygenupdater.deviceinformation.DeviceInformationFragment
import com.arjanvlek.oxygenupdater.internal.KotlinCallback
import com.arjanvlek.oxygenupdater.internal.Utils
import com.arjanvlek.oxygenupdater.models.Banner
import com.arjanvlek.oxygenupdater.models.DeviceOsSpec
import com.arjanvlek.oxygenupdater.models.DeviceRequestFilter
import com.arjanvlek.oxygenupdater.news.NewsFragment
import com.arjanvlek.oxygenupdater.notifications.Dialogs
import com.arjanvlek.oxygenupdater.notifications.MessageDialog
import com.arjanvlek.oxygenupdater.notifications.NotificationTopicSubscriber.subscribe
import com.arjanvlek.oxygenupdater.settings.SettingsActivity.Companion.SKU_AD_FREE
import com.arjanvlek.oxygenupdater.settings.SettingsManager
import com.arjanvlek.oxygenupdater.settings.SettingsManager.Companion.PROPERTY_AD_FREE
import com.arjanvlek.oxygenupdater.settings.adFreeVersion.util.IabHelper
import com.arjanvlek.oxygenupdater.settings.adFreeVersion.util.IabHelper.IabAsyncInProgressException
import com.arjanvlek.oxygenupdater.settings.adFreeVersion.util.PK1
import com.arjanvlek.oxygenupdater.settings.adFreeVersion.util.PK2
import com.arjanvlek.oxygenupdater.updateinformation.ServerMessageBar
import com.arjanvlek.oxygenupdater.updateinformation.UpdateInformationFragment
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.InterstitialAd
import com.google.android.gms.ads.MobileAds
import com.google.android.material.appbar.AppBarLayout
import kotlinx.android.synthetic.main.activity_main.*
import org.joda.time.LocalDateTime
import java.util.*
import kotlin.math.abs

class MainActivity : AppCompatActivity(), Toolbar.OnMenuItemClickListener {

    private lateinit var viewPager: ViewPager
    private lateinit var settingsManager: SettingsManager
    lateinit var activityLauncher: ActivityLauncher
        private set

    private var serverMessageBars = ArrayList<ServerMessageBar>()
    private var newsAd: InterstitialAd? = null
    private var downloadPermissionCallback: KotlinCallback<Boolean>? = null
    private var adView: AdView? = null

    var deviceOsSpec: DeviceOsSpec? = null
        private set

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        settingsManager = SettingsManager(this)

        // App version 2.4.6: Migrated old setting Show if system is up to date (default: ON) to Advanced mode (default: OFF).
        if (settingsManager.containsPreference(SettingsManager.PROPERTY_SHOW_IF_SYSTEM_IS_UP_TO_DATE)) {
            settingsManager.savePreference(SettingsManager.PROPERTY_ADVANCED_MODE, !settingsManager.getPreference(SettingsManager.PROPERTY_SHOW_IF_SYSTEM_IS_UP_TO_DATE, true))
            settingsManager.deletePreference(SettingsManager.PROPERTY_SHOW_IF_SYSTEM_IS_UP_TO_DATE)
        }

        val application = application as ApplicationData

        // Supported device check
        application.serverConnector!!.getDevices(DeviceRequestFilter.ALL) { result ->
            deviceOsSpec = Utils.checkDeviceOsSpec(application.systemVersionProperties!!, result)

            val showDeviceWarningDialog = !settingsManager.getPreference(SettingsManager.PROPERTY_IGNORE_UNSUPPORTED_DEVICE_WARNINGS, false)

            if (showDeviceWarningDialog && !deviceOsSpec!!.isDeviceOsSpecSupported) {
                displayUnsupportedDeviceOsSpecMessage(deviceOsSpec!!)
            }
        }

        toolbar.setOnMenuItemClickListener(this)

        setupViewPager()

        activityLauncher = ActivityLauncher(this)

        // Set start page to Update Information Screen (middle page)
        try {
            var startPage = PAGE_UPDATE_INFORMATION
            val extras = intent?.extras

            if (extras?.containsKey(INTENT_START_PAGE) == true) {
                startPage = extras.getInt(INTENT_START_PAGE)
            }

            viewPager.currentItem = startPage
        } catch (ignored: IndexOutOfBoundsException) {
            // no-op
        }

        setupAds()

        // Support functions for Android 8.0 "Oreo" and up.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createPushNotificationChannel()
            createProgressNotificationChannel()
        }

        // Remove "long" download ID in favor of "int" id
        try {
            // Do not remove this int assignment, even though the IDE warns it's unused.
            // getPreference method has a generic signature, we need to force its return type to be an int,
            // otherwise it triggers a ClassCastException (which occurs when coming from older app versions)
            // whilst not assigning it converts it to a long
            @Suppress("UNUSED_VARIABLE")
            val downloadId = settingsManager.getPreference(SettingsManager.PROPERTY_DOWNLOAD_ID, -1)
        } catch (e: ClassCastException) {
            settingsManager.deletePreference(SettingsManager.PROPERTY_DOWNLOAD_ID)
        }

        // Offer contribution to users from app versions below 2.4.0
        if (!settingsManager.containsPreference(SettingsManager.PROPERTY_CONTRIBUTE) && settingsManager.containsPreference(SettingsManager.PROPERTY_SETUP_DONE)) {
            activityLauncher.Contribute()
        }
    }

    public override fun onStart() {
        super.onStart()

        // Mark the welcome tutorial as finished if the user is moving from older app version.
        // This is checked by either having stored update information for offline viewing,
        // or if the last update checked date is set (if user always had up to date system and never viewed update information before)
        if (!settingsManager.getPreference(SettingsManager.PROPERTY_SETUP_DONE, false)
            && (settingsManager.checkIfOfflineUpdateDataIsAvailable() || settingsManager.containsPreference(SettingsManager.PROPERTY_UPDATE_CHECKED_DATE))
        ) {
            settingsManager.savePreference(SettingsManager.PROPERTY_SETUP_DONE, true)
        }

        // Show the welcome tutorial if the app needs to be set up
        if (!settingsManager.getPreference(SettingsManager.PROPERTY_SETUP_DONE, false)) {
            if (Utils.checkNetworkConnection(applicationContext)) {
                activityLauncher.Tutorial()
            } else {
                showNetworkError()
            }
        } else {
            if (!settingsManager.containsPreference(SettingsManager.PROPERTY_NOTIFICATION_TOPIC)) {
                subscribe((application as ApplicationData))
            }
        }

        val application = application as ApplicationData
        val online = Utils.checkNetworkConnection(application)
        application.serverConnector?.getInAppMessages(online, { displayServerMessageBars(it) }) { error ->
            when (error) {
                ApplicationData.SERVER_MAINTENANCE_ERROR -> Dialogs.showServerMaintenanceError(this)
                ApplicationData.APP_OUTDATED_ERROR -> Dialogs.showAppOutdatedError(this)
            }
        }
    }

    /*
      -------------- METHODS FOR DISPLAYING DATA ON THE FRAGMENT -------------------
     */
    private fun addServerMessageBar(view: ServerMessageBar) {
        // Add the message to the update information screen.
        // Set the layout params based on the view count.
        // First view should go below the app update message bar (if visible)
        // Consecutive views should go below their parent / previous view.
        val params = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)

        // position each bar below the previous one
        if (serverMessageBars.isNotEmpty()) {
            // params.topToBottom = serverMessageBars[serverMessageBars.size - 1].id
        }

        view.id = View.generateViewId()
        view.visibility = GONE
        serverMessageLayout.addView(view, params)
        serverMessageBars.add(view)
    }

    private fun deleteAllServerMessageBars() {
        serverMessageBars.forEach {
            serverMessageLayout.removeView(it)
        }

        serverMessageBars = ArrayList()
    }

    private fun displayServerMessageBars(banners: List<Banner>) {
        if (isFinishing) {
            return
        }

        deleteAllServerMessageBars()

        val createdServerMessageBars = ArrayList<ServerMessageBar>()

        banners.forEach {
            val bar = ServerMessageBar(this)
            val backgroundBar = bar.backgroundBar
            val textView = bar.textView

            backgroundBar.setBackgroundColor(it.getColor(this))
            textView.text = it.getBannerText(this)

            if (it.getBannerText(this) is Spanned) {
                textView.movementMethod = LinkMovementMethod.getInstance()
            }

            addServerMessageBar(bar)
            createdServerMessageBars.add(bar)
        }

        // Position the app UI  to be below the last added server message bar
        if (createdServerMessageBars.isNotEmpty()) {
            serverMessageLayout.apply {
                // visibility = VISIBLE

                setOnClickListener {
                    children.forEach {
                        if (it is ServerMessageBar) {
                            it.visibility = if (it.visibility == VISIBLE) GONE else VISIBLE
                        } else if (it is TextView) {
                            it.text = if (it.text == "Tap here to view notices") "Tap here to collapse" else "Tap here to view notices"
                        }
                    }
                }
            }
        } else {
            serverMessageLayout.visibility = GONE
        }

        serverMessageBars = createdServerMessageBars
    }

    /**
     * Handles toolbar menu clicks
     *
     * @param item the menu item
     *
     * @return true if clicked
     */
    override fun onMenuItemClick(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> {
                activityLauncher.Settings()
                true
            }
            R.id.action_about -> {
                activityLauncher.About()
                true
            }
            R.id.action_help -> {
                activityLauncher.Help()
                true
            }
            R.id.action_faq -> {
                activityLauncher.FAQ()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setupViewPager() {
        viewpager.apply {
            viewPager = this

            offscreenPageLimit = 2
            adapter = SectionsPagerAdapter(supportFragmentManager)
            tabs.setupWithViewPager(this)

            appBar.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { appBarLayout, verticalOffset ->
                val params = this.layoutParams as CoordinatorLayout.LayoutParams
                params.bottomMargin = appBarLayout.totalScrollRange - abs(verticalOffset)

                this.layoutParams = params
            })
        }
    }

    /**
     * Checks for Play Services and initialises [MobileAds] if found
     */
    private fun setupAds() {
        val application = application as ApplicationData

        if (application.checkPlayServices(this, false)) {
            MobileAds.initialize(this, getString(R.string.advertising_app_id))
        } else {
            Toast.makeText(application, getString(R.string.notification_no_notification_support), Toast.LENGTH_LONG).show()
        }

        if (!settingsManager.getPreference(PROPERTY_AD_FREE, false)) {
            InterstitialAd(this).apply {
                newsAd = this

                adUnitId = getString(R.string.advertising_interstitial_unit_id)
                loadAd(buildAdRequest())
            }
        }

        adView = updateInformationAdView
        checkAdSupportStatus { adsAreSupported ->
            if (adsAreSupported) {
                // need to add spacing between ViewPager contents and the AdView to avoid overlapping the last item
                // Since the AdView's size is BANNER, bottom padding should be 50dp
                viewPager.setPadding(0, 0, 0, Utils.dpToPx(this, 66f).toInt())

                adView?.visibility = VISIBLE
                showAds()
            } else {
                // reset viewPager padding
                viewPager.setPadding(0, 0, 0, 0)

                adView?.visibility = GONE
            }
        }
    }

    public override fun onResume() {
        super.onResume()
        adView?.resume()
    }

    public override fun onPause() {
        super.onPause()
        adView?.pause()
    }

    public override fun onDestroy() {
        super.onDestroy()
        adView?.destroy()
    }

    /**
     * Checks if ads should be shown
     *
     * @param callback the callback that receives the result: false if user has purchased ad-free, true otherwise
     */
    private fun checkAdSupportStatus(callback: KotlinCallback<Boolean>) {
        val helper = IabHelper(this, PK1.A + "/" + PK2.B)

        helper.startSetup { setupResult ->
            if (!setupResult.success) {
                // Failed to setup IAB, so we might be offline or the device does not support IAB. Return the last stored value of the ad-free status.
                callback.invoke(!settingsManager.getPreference(PROPERTY_AD_FREE, false))
                return@startSetup
            }

            try {
                helper.queryInventoryAsync(true, listOf(SKU_AD_FREE), null) { queryResult, inventory ->
                    if (!queryResult.success) {
                        // Failed to check inventory, so we might be offline. Return the last stored value of the ad-free status.
                        callback.invoke(!settingsManager.getPreference(PROPERTY_AD_FREE, false))
                    } else {
                        if (inventory != null && inventory.hasPurchase(SKU_AD_FREE)) {
                            // User has bought the upgrade. Save this to the app's settings and return that ads may not be shown.
                            settingsManager.savePreference(PROPERTY_AD_FREE, true)
                            callback.invoke(false)
                        } else {
                            // User has not bought the item and we're online, so ads are definitely supported
                            callback.invoke(true)
                        }
                    }
                }
            } catch (e: IabAsyncInProgressException) {
                // A check is already in progress, so wait 3 secs and try to check again.
                Handler().postDelayed({ checkAdSupportStatus(callback) }, 3000)
            }
        }
    }

    /**
     * Loads an ad
     */
    private fun showAds() {
        adView!!.loadAd(buildAdRequest())
    }

    fun displayUnsupportedDeviceOsSpecMessage(deviceOsSpec: DeviceOsSpec) {
        // Do not show dialog if app was already exited upon receiving of devices from the server.
        if (isFinishing) {
            return
        }

        val resourceId = when (deviceOsSpec) {
            DeviceOsSpec.CARRIER_EXCLUSIVE_OXYGEN_OS -> R.string.carrier_exclusive_device_warning_message
            DeviceOsSpec.UNSUPPORTED_OXYGEN_OS -> R.string.unsupported_device_warning_message
            DeviceOsSpec.UNSUPPORTED_OS -> R.string.unsupported_os_warning_message
            else -> R.string.unsupported_os_warning_message
        }

        val checkBoxView = View.inflate(this@MainActivity, R.layout.message_dialog_checkbox, null)

        AlertDialog.Builder(this)
            .setView(checkBoxView)
            .setTitle(getString(R.string.unsupported_device_warning_title))
            .setMessage(getString(resourceId))
            .setPositiveButton(getString(R.string.download_error_close)) { dialog, _ ->
                val checkbox = checkBoxView.findViewById<CheckBox>(R.id.unsupported_device_warning_checkbox)
                settingsManager.savePreference(SettingsManager.PROPERTY_IGNORE_UNSUPPORTED_DEVICE_WARNINGS, checkbox.isChecked)
                dialog.dismiss()
            }
            .show()
    }

    private fun showNetworkError() {
        if (!isFinishing) {
            MessageDialog(
                this,
                title = getString(R.string.error_app_requires_network_connection),
                message = getString(R.string.error_app_requires_network_connection_message),
                negativeButtonText = getString(R.string.download_error_close),
                cancellable = false
            ).show()
        }
    }

    fun getNewsAd(): InterstitialAd? {
        return when {
            newsAd != null -> newsAd
            mayShowNewsAd() -> {
                InterstitialAd(this).apply {
                    adUnitId = getString(R.string.advertising_interstitial_unit_id)
                    loadAd(buildAdRequest())

                    newsAd = this
                    newsAd
                }
            }
            else -> null
        }
    }

    fun mayShowNewsAd(): Boolean {
        return (!settingsManager.getPreference(PROPERTY_AD_FREE, false)
                && LocalDateTime.parse(settingsManager.getPreference(SettingsManager.PROPERTY_LAST_NEWS_AD_SHOWN, "1970-01-01T00:00:00.000"))
            .isBefore(LocalDateTime.now().minusMinutes(5)))
    }

    fun requestDownloadPermissions(callback: KotlinCallback<Boolean>) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            downloadPermissionCallback = callback
            requestPermissions(arrayOf(DOWNLOAD_FILE_PERMISSION, VERIFY_FILE_PERMISSION), PERMISSION_REQUEST_CODE)
        }
    }

    override fun onRequestPermissionsResult(permsRequestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (permsRequestCode == PERMISSION_REQUEST_CODE && grantResults.isNotEmpty()) {
            downloadPermissionCallback?.invoke(grantResults[0] == PERMISSION_GRANTED)
        }
    }

    // Android 6.0 Run-time permissions
    fun hasDownloadPermissions(): Boolean {
        return (ContextCompat.checkSelfPermission(application, VERIFY_FILE_PERMISSION) == PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(application, DOWNLOAD_FILE_PERMISSION) == PERMISSION_GRANTED)
    }

    @TargetApi(26)
    private fun createPushNotificationChannel() {
        // The id of the channel.
        val id = ApplicationData.PUSH_NOTIFICATION_CHANNEL_ID

        // The user-visible name of the channel.
        val name: CharSequence = getString(R.string.push_notification_channel_name)

        // The user-visible description of the channel.
        NotificationChannel(id, name, NotificationManager.IMPORTANCE_HIGH).apply {
            // Configure the notification channel.
            description = getString(R.string.push_notification_channel_description)
            enableLights(true)

            // Sets the notification light color for notifications posted to this
            // channel, if the device supports this feature.
            lightColor = Color.RED
            enableVibration(true)
            vibrationPattern = longArrayOf(100, 200, 300, 400, 500, 400, 300, 200, 400)

            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
            notificationManager?.createNotificationChannel(this)
        }
    }

    @TargetApi(26)
    private fun createProgressNotificationChannel() {
        // The id of the channel.
        val id = ApplicationData.PROGRESS_NOTIFICATION_CHANNEL_ID

        // The user-visible name of the channel.
        val name = getString(R.string.progress_notification_channel_name)

        // The user-visible description of the channel.
        NotificationChannel(id, name, NotificationManager.IMPORTANCE_LOW).apply {
            // Configure the notification channel.
            description = getString(R.string.progress_notification_channel_description)

            enableLights(false)
            enableVibration(false)

            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
            notificationManager?.createNotificationChannel(this)
        }
    }
    // Android 8.0 Notification Channels
    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to one of the sections/tabs/pages.
     */
    inner class SectionsPagerAdapter internal constructor(fragmentManager: FragmentManager) : FragmentPagerAdapter(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
        override fun getItem(position: Int): Fragment {
            // getItem is called to instantiate the fragment for the given page.
            return when (position) {
                PAGE_NEWS -> NewsFragment()
                PAGE_UPDATE_INFORMATION -> UpdateInformationFragment()
                PAGE_DEVICE_INFORMATION -> DeviceInformationFragment()
                else -> TODO()
            }
        }

        override fun getCount(): Int {
            // Show 3 total pages.
            return 3
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return when (position) {
                PAGE_NEWS -> getString(R.string.news)
                PAGE_UPDATE_INFORMATION -> getString(R.string.update_information_header_short)
                PAGE_DEVICE_INFORMATION -> getString(R.string.device_information_header_short)
                else -> null
            }
        }
    }

    interface DeviceOsSpecCheckedListener {
        fun onDeviceOsSpecChecked(deviceOsSpec: DeviceOsSpec)
    }

    companion object {
        const val PAGE_NEWS = 0
        const val PAGE_UPDATE_INFORMATION = 1
        const val PAGE_DEVICE_INFORMATION = 2
        // Permissions constants
        const val PERMISSION_REQUEST_CODE = 200
        const val DOWNLOAD_FILE_PERMISSION = "android.permission.WRITE_EXTERNAL_STORAGE"
        const val VERIFY_FILE_PERMISSION = "android.permission.READ_EXTERNAL_STORAGE"
        const val INTENT_START_PAGE = "start_page"
    }
}
